### PR TITLE
Translate ArrayProxy to pattern matching when supported

### PR DESCRIPTION
**Pull request by Jimx-**
Current the value compiler translates ArrayProxy into if-elif trees which can cause the compiler to crash due to deep recursion (#359). This patch instead translates them into pattern matching when it is supported (Python >= 3.10) to avoid this problem.

### CODE REVIEW

**Comment by whitequark** [comment]
This is an interesting solution--I'll need to look at it a bit closer.

What's performance like?

**Comment by Jimx-** [comment]
Pattern matching seems to be slightly faster than if-elif for compiling
```python
m = Module()
mem = Memory(width=32, depth=0xb80)
m.submodules.rport = mem.read_port()
m.submodules.wport = mem.write_port()
sim = Simulator(m)
sim.run()
```
Pattern matching:
`python test.py  0.28s user 0.02s system 99% cpu 0.299 total`

if-elif:
`python test.py  0.36s user 0.02s system 99% cpu 0.380 total`

If-elif crashes for depth over 0xc00 with the default recursion limit

Patter matching with depth 0x4000:
`python test.py  1.48s user 0.10s system 99% cpu 1.581 total`

**Comment by whitequark** [comment]
What about runtime performance?

**Comment by Jimx-** [comment]
They are very close

```python
    def process():
        for i in range(depth):
            yield dut.waddr.eq(i)
            yield dut.wdata.eq(i)
            yield dut.we.eq(1)
            yield

    sim.add_sync_process(process)
    start = time.time()
    sim.run()
    end = time.time()
    print(end - start)
```
It took both of them about 1.5s to finish

File: amaranth/sim/_pyrtl.py [hunk 1]
```diff
@@ -224,20 +225,33 @@ def on_Repl(self, value):
         return f"0"
 
     def on_ArrayProxy(self, value):
+        use_pm = sys.version_info >= (3, 10)
```
Comment (whitequark) [inline amaranth/sim/_pyrtl.py:228]: Can you make this a private module-level constant? Something like `_USE_PATTERN_MATCHING = (sys.version_info >= (3, 10))` after the `__all__` definition.

**Review by whitequark — COMMENTED** [review]
Looks good to me; I had another solution planned but there's no reason to not merge this in the interim.

**Comment by whitequark** [comment]
Could you rebase on top of `main`, please? The CI should work afterwards.

**Review by whitequark — APPROVED** [review]

**Comment by whitequark** [comment]
Thank you!

**Pull request merged**